libs/ml/examples/bert_save_load_pretrained_model_example/main.cpp

```diff
@@ -0,0 +1,527 @@
+
+  // add layers as well as weights
+  std::string              layer_output = norm_embed;
+  std::vector<std::string> encoder_outputs;
+  encoder_outputs.emplace_back(layer_output);
+  FETCH_UNUSED(n_encoder_layers);
+  for (SizeType i = 0u; i < n_encoder_layers; i++)
+  {
+    // create the encoding layer first
+    layer_output = g.template AddNode<fetch::ml::layers::SelfAttentionEncoder<TensorType>>(
+        "SelfAttentionEncoder_No_" + std::to_string(i), {layer_output, mask}, n_heads, model_dims,
+        ff_dims, dropout_keep_prob, dropout_keep_prob, dropout_keep_prob, epsilon);
+
+    // store layer output
+    encoder_outputs.emplace_back(layer_output);
+
+    // get state dict for this layer
+    state_dict =
+        std::dynamic_pointer_cast<GraphType>(g.GetNode(layer_output)->GetOp())->StateDict();
+
+    // get file path prefix
+    std::string file_prefix = file_path + "bert_encoder_layer_" + std::to_string(i) + "_";
+
+    // put weights in 2 layer norms
+    PutWeightInLayerNorm(state_dict, model_dims, file_prefix + "attention_output_LayerNorm_gamma",
+                         file_prefix + "attention_output_LayerNorm_beta",
+                         "SelfAttentionEncoder_Attention_Residual_LayerNorm_LayerNorm_Gamma",
+                         "SelfAttentionEncoder_Attention_Residual_LayerNorm_LayerNorm_Beta");
+    PutWeightInLayerNorm(state_dict, model_dims, file_prefix + "output_LayerNorm_gamma",
+                         file_prefix + "output_LayerNorm_beta",
+                         "SelfAttentionEncoder_Feedforward_Residual_LayerNorm_LayerNorm_Gamma",
+                         "SelfAttentionEncoder_Feedforward_Residual_LayerNorm_LayerNorm_Beta");
+
+    // put weights in ff block and attention linear conversion part
+    PutWeightInFullyConnected(
+        state_dict, model_dims, ff_dims, file_prefix + "intermediate_dense_weight",
+        file_prefix + "intermediate_dense_bias",
+        "SelfAttentionEncoder_Feedforward_Feedforward_No_1_TimeDistributed_FullyConnected_Weights",
+        "SelfAttentionEncoder_Feedforward_Feedforward_No_1_TimeDistributed_FullyConnected_Bias");
+    PutWeightInFullyConnected(
+        state_dict, ff_dims, model_dims, file_prefix + "output_dense_weight",
+        file_prefix + "output_dense_bias",
+        "SelfAttentionEncoder_Feedforward_Feedforward_No_2_TimeDistributed_FullyConnected_Weights",
+        "SelfAttentionEncoder_Feedforward_Feedforward_No_2_TimeDistributed_FullyConnected_Bias");
+    PutWeightInFullyConnected(state_dict, model_dims, model_dims,
+                              file_prefix + "attention_output_dense_weight",
+                              file_prefix + "attention_output_dense_bias",
+                              "SelfAttentionEncoder_Multihead_Attention_MultiheadAttention_"
+                              "Final_Transformation_TimeDistributed_FullyConnected_Weights",
+                              "SelfAttentionEncoder_Multihead_Attention_MultiheadAttention_"
```

Comment:
these all seem problematic because if we change any of the default naming convention for these ops these will all break. is there no way to set these names dynamically based out the output from graph.AddNode()? I guess this is a problem with how we manage nested layers at the moment. I guess we should store the layer names as variables or otherwise make them accessible dynamically

Reply:
well it is quite troublesome to change. and actually it is quite difficult to implement cuz the add node process is hidden in subgraph.